clients/src/main/java/org/apache/kafka/clients/producer/internals/ProducerInterceptors.java

```diff
@@ -0,0 +1,103 @@
+    private final List<ProducerInterceptor<K, V>> interceptors;
+
+    public ProducerInterceptors(List<ProducerInterceptor<K, V>> interceptors) {
+        this.interceptors = interceptors;
+    }
+
+    /**
+     * This is called when client sends the record to KafkaProducer, before key and value gets serialized.
+     * The method calls {@link ProducerInterceptor#onSend(ProducerRecord)} method. ProducerRecord
+     * returned from the first interceptor's onSend() is passed to the second interceptor onSend(), and so on in the
+     * interceptor chain. The record returned from the last interceptor is returned from this method.
+     *
+     * This method does not throw exceptions. Exceptions thrown by any of interceptor methods are caught and ignored.
+     * If an interceptor in the middle of the chain, that normally modifies the record, throws an exception,
+     * the next interceptor in the chain will be called with a record returned by the previous interceptor that did not
+     * throw an exception.
+     *
+     * @param record the record from client
+     * @return producer record to send to topic/partition
+     */
+    public ProducerRecord<K, V> onSend(ProducerRecord<K, V> record) {
+        ProducerRecord<K, V> interceptRecord = record;
+        for (ProducerInterceptor<K, V> interceptor: this.interceptors) {
+            try {
+                interceptRecord = interceptor.onSend(interceptRecord);
+            } catch (Throwable t) {
+                // do not propagate interceptor exception, ignore and continue calling other interceptors
+                log.warn("Error executing interceptor onSend callback for topic: {}, partition: {}, with error: {}",
+                        record.topic(), record.partition(), t);
+            }
+        }
+        return interceptRecord;
+    }
+
+    /**
+     * This method is called when the record sent to the server has been acknowledged, or when sending the record fails before
+     * it gets sent to the server. This method calls {@link ProducerInterceptor#onAcknowledgement(RecordMetadata, Exception)}
+     * method for each interceptor.
+     *
+     * This method does not throw exceptions. Exceptions thrown by any of interceptor methods are caught and ignored.
+     *
+     * @param metadata The metadata for the record that was sent (i.e. the partition and offset). Null if an error occurred.
+     * @param exception The exception thrown during processing of this record. Null if no error occurred.
+     */
+    public void onAcknowledgement(RecordMetadata metadata, Exception exception) {
+        for (ProducerInterceptor<K, V> interceptor: this.interceptors) {
+            try {
+                interceptor.onAcknowledgement(metadata, exception);
+            } catch (Throwable t) {
+                // do not propagate interceptor exceptions, just ignore
+                log.warn("Error executing interceptor onAcknowledgement callback: {}", t);
```

Comment:
You can remove `: {}` as we are not passing any args anymore, that is, we are calling the second method instead of the first:

`public void warn(String format, Object arg);`
`public void warn(String msg, Throwable t);`